src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1890,3 +1955,58 @@ fn is_primitive(path_str: &str, ns: Namespace) -> Option<(&'static str, Res)> {
 fn primitive_impl(cx: &DocContext<'_>, path_str: &str) -> Option<&'static SmallVec<[DefId; 4]>> {
     Some(PrimitiveType::from_symbol(Symbol::intern(path_str))?.impls(cx.tcx))
 }
+
+fn strip_generics_from_path(path_str: &str) -> Result<String, ResolutionFailure<'static>> {
+    let mut stripped_segments = vec![];
+
+    for segment in path_str.split("::") {
```

Comment:
Do we need to actually parse `::` as splits? Can we not just split on the first `<`, strip a `::` from the first segment, and ensure the second ends with `>`?

Splitting on `::` will mean `::` isn't allowed in angle brackets, and it will be slower.

Reply:
Hmm. How would that work with something like `Vec::<Box<T>>::new()`?

Reply:
Or are you suggesting that maybe we don't support that?

Reply:
Related: I think you'll currently give weird errors for `Vec::<std::error::Error>::new`.

Reply:
Yeah, looks like this:

```
error: unresolved link to `Vec::<std::error::Error>::new`
  --> $DIR/intra-link-malformed-generics.rs:12:6
   |
LL | //! [Vec::<std::error::Error>::new]
   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unbalanced angle brackets
```

Reply:
I guess it's time to reimplement the Rust parser ;)

Reply:
Hmm, doing (relatively) simple parsing might fix it.

Reply:
I'd prefer to reuse the compiler"s parsing if possible ... This is getting somewhat complicated.

@manishearth do you have ideas for doing that?

Reply:
Maybe we could use `rustc_parse` somehow.

Reply:
> Or are you suggesting that maybe we don't support that?

I would rather not support that, yes.

Reply:
Either way, it's not as easy as splitting on colons because `Foo<bar::Baz>` exists. I don't want to reuse the compiler's parsing because it's relatively heavy and not efficient for what we'd like

Reply:
Basically, for intra doc links to work we need to be able to quick-reject URLs, and parse paths quickly, and this is a pretty slow parse step.

Reply:
> I don't want to reuse the compiler's parsing because it's relatively heavy and not efficient for what we'd like

I'm not sure why you say that. All this is only running if we saw a `<` anyway, so in the common case we never have to go through the compiler, and if there is a `<` I don't see why anything we do would be faster than the compiler.

> > Or are you suggesting that maybe we don't support [... `Vec::<T>::new` ...]?

> I would rather not support that, yes.

I'm ok with that for now, but it will have to be implemented for #74563 at some point. I really don't want to rewrite the parsing and have to keep it in sync with the compiler.

Reply:
I don't know how hard it is to invoke the parser, but I think it might be better to just write a lightweight parser like Manishearth suggested. It should only need to understand basic paths and generics, and it would probably be pretty small. What do you think @jyn514?

Reply:
So I tried writing this and got this far:
```rust
fn parse(cx: &DocContext<'_>, source: String) {
    use rustc_span::FileName;
    use rustc_session::parse::ParseSess;
    use rustc_parse::parser::Parser;

    let filename = FileName::Custom("intra-doc link".into());
    let parse_sess = &cx.tcx.sess.parse_sess;
    let tokens = rustc_parse::parse_stream_from_source_str(filename, source, parse_sess, None);
    let mut parser = Parser::new(parse_sess, tokens, false, None);
    match parser.parse_ty() {

    }
}
```
and then I realized this is on the wrong track because `parse_qpath` is private and you have to use either `parse_ty` or `parse_expr`. So ... I don't know. I'd really prefer not to rewrite a parser from scratch :/

Reply:
I think the parser will only have to understand `::` and angle brackets, and can treat everything else as text, so I don't think it will be too complex. I'll try implementing something simple.